### PR TITLE
Extending join fuzz tests to support join filtering

### LINKED ISSUE #10659: SortMergeJoin: Add fuzz tests for SMJ when the JoinFilter is set
### Is your feature request related to a problem or challenge?

SMJ now is in experimental state, and the test coverage can be improved. Because of poor test coverage it was found multiple bugs manually, with tests it would be much easier

### Describe the solution you'd like

I would like to expand fuzz test coverage for SMJ in `join_fuzz.rs` to support scenarios when the JoinFilter is set.
https://github.com/apache/datafusion/blob/8bedecc00b2f1f04d7b1a907152ce0d19b7046a5/datafusion/core/tests/fuzz_cases/join_fuzz.rs#L179

Currently JoinFilter is hardcoded as None for SMJ and would be great to start work for adding filter test cases

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edmondop**
## Which issue does this PR close?

Closes
#10659

### CODE REVIEW

**Comment by comphead** [comment]
Thanks @edmondop I'm planning to read it through soon

**Comment by comphead** [comment]
> Yes I was stuck because one test doesn't pass, will need to investigate unless you have some hints to share

If you referring to CI test failed, it is 
```
  left: (0, "+----+----+-------------+------------+")
 right: (0, "+----+----+-------------+-------------+")
```
which should probably be straightforward to fix, we can check row counts first?

**Comment by edmondop** [comment]
> > Yes I was stuck because one test doesn't pass, will need to investigate unless you have some hints to share
> 
> If you referring to CI test failed, it is
> 
> ```
>   left: (0, "+----+----+-------------+------------+")
>  right: (0, "+----+----+-------------+-------------+")
> ```
> 
> which should probably be straightforward to fix, we can check row counts first?

Thanks for the tip, checking for the row count effectively show that the right_join filtered return a different number of rows when using SMJ and HashJoin. What could be the root cause?

I have tried to modify the column used for the filter un-commenting the section commented below, but this breaks all the other test cases too

```rust
fn less_than_10_join_filter(schema1: Arc<Schema>, _schema2: Arc<Schema>) -> JoinFilter {
    let less_than_100 = Arc::new(BinaryExpr::new(
        Arc::new(Column::new("a", 0)),
        Operator::Lt,
        Arc::new(Literal::new(ScalarValue::from(100))),
    )) as _;
    let column_indices = vec![
        ColumnIndex {
            index: 0,
            side: JoinSide::Left,
        },
        // ColumnIndex {
        //     index: 0,
        //     side: JoinSide::Right,
        // },
    ];
    let intermediate_schema =
        Schema::new(vec![schema1.field_with_name("a").unwrap().to_owned()]);

    JoinFilter::new(less_than_100, column_indices, intermediate_schema)
}
```

**Comment by edmondop** [comment]
> Details

These are the two tests that fails

```
failures:
    fuzz_cases::join_fuzz::test_anti_join_1k_filtered
    fuzz_cases::join_fuzz::test_right_join_1k_filtered
```


```
---- fuzz_cases::join_fuzz::test_anti_join_1k_filtered stdout ----
thread 'fuzz_cases::join_fuzz::test_anti_join_1k_filtered' panicked at datafusion/core/tests/fuzz_cases/join_fuzz.rs:404:13:
assertion `left == right` failed: SortMergeJoinExec and HashJoinExec produced different row counts
  left: 5
 right: 902
```

right join passes on my machine, I think I need to rebase, but also it seems easier to solve

```
---- fuzz_cases::join_fuzz::test_right_join_1k_filtered stdout ----
thread 'fuzz_cases::join_fuzz::test_right_join_1k_filtered' panicked at datafusion/core/tests/fuzz_cases/join_fuzz.rs:377:70:
called `Result::unwrap()` on an `Err` value: ArrowError(InvalidArgumentError("Column 'a' is declared as non-nullable but contains null values"), None)
stack backtrace:
```

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 1]
```diff
@@ -40,92 +45,203 @@ use test_utils::stagger_batch_with_seed;
...
         make_staggered_batches(1000),
         make_staggered_batches(1000),
         JoinType::Left,
+        Some(Box::new(less_than_10_join_filter)),
     )
+    .run_test()
     .await
 }
 
 #[tokio::test]
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:122]:
```suggestion
// Add support for RightJoins
#[ignore]
#[tokio::test]
```

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 2]
```diff
@@ -40,92 +45,203 @@ use test_utils::stagger_batch_with_seed;
...
+        make_staggered_batches(1000),
+        make_staggered_batches(1000),
+        JoinType::LeftAnti,
+        None,
+    )
+    .run_test()
+    .await
+}
+
+#[tokio::test]
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:205]:
```suggestion
// Test failed for now. 
#[ignore]
#[tokio::test]
```

**Comment by comphead** [comment]
@edmondop I think it is a solid start already, lets comment this tests for now and address them separately

**Comment by comphead** [comment]
I took the liberty to ignore tests

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 3]
```diff
@@ -40,92 +45,207 @@ use test_utils::stagger_batch_with_seed;
...
         make_staggered_batches(1000),
         JoinType::Left,
+        Some(Box::new(less_than_10_join_filter)),
     )
+    .run_test()
     .await
 }
 
+// Add support for RightJoins
+#[ignore]
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:123]:
```suggestion
```

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 4]
```diff
@@ -40,92 +45,207 @@ use test_utils::stagger_batch_with_seed;
...
         make_staggered_batches(1000),
         make_staggered_batches(1000),
         JoinType::Left,
+        Some(Box::new(less_than_10_join_filter)),
     )
+    .run_test()
     .await
 }
 
+// Add support for RightJoins
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:122]:
```suggestion
```

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 5]
```diff
@@ -40,92 +45,207 @@ use test_utils::stagger_batch_with_seed;
...
+    JoinFuzzTestCase::new(
+        make_staggered_batches(1000),
+        make_staggered_batches(1000),
+        JoinType::Right,
+        None,
+    )
+    .run_test()
+    .await
+}
+#[tokio::test]
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:135]:
```suggestion
// Add support for Right filtered joins
#[ignored]
#[tokio::test]
```

File: datafusion/core/tests/fuzz_cases/join_fuzz.rs [hunk 6]
```diff
@@ -40,92 +45,207 @@ use test_utils::stagger_batch_with_seed;
...
+        make_staggered_batches(1000),
+        make_staggered_batches(1000),
+        JoinType::Right,
+        None,
+    )
+    .run_test()
+    .await
+}
+// Add support for Right filtered joins
+#[ignored]
```
Comment (comphead) [inline datafusion/core/tests/fuzz_cases/join_fuzz.rs:134]:
```suggestion
#[ignore]
```

**Comment by comphead** [comment]
Right filtered join fuzz tests failures probably related to https://github.com/apache/datafusion/issues/10882

**Comment by edmondop** [comment]
Fixed formatting @comphead all the checks are passing now

**Review by comphead — APPROVED** [review]
lgtm well done @edmondop 
We plan to address anti and right filtered joins separately and then uncomment the test

**Comment by ozankabak** [comment]
FYI, CI is failing on main after merge.

**Comment by edmondop** [comment]
> FYI, CI is failing on main after merge.

This should fix it https://github.com/apache/datafusion/pull/10887. I also filed #10886 cc @comphead

**Pull request merged**